barretenberg/cpp/src/barretenberg/world_state/world_state.cpp

```diff
@@ -384,19 +394,27 @@ bool WorldState::sync_block(StateReference& block_state_ref,
     }
 
     {
+        // insert public writes in batches so that we can have different transactions modifying the same slot in the
+        // same L2 block
         auto& wrapper = std::get<TreeWithStore<PublicDataTree>>(fork->_trees.at(MerkleTreeId::PUBLIC_DATA_TREE));
-        std::vector<PublicDataLeafValue> leaves;
-        size_t total_leaves = 0;
-        for (const auto& batch : public_writes) {
-            total_leaves += batch.size();
-        }
+        std::shared_ptr<size_t> current_batch = std::make_shared<size_t>(0);
+        std::shared_ptr<PublicDataTree::AddCompletionCallback> completion =
+            std::make_shared<PublicDataTree::AddCompletionCallback>();
+
+        *completion = [&decr, &public_writes, &wrapper, completion, current_batch](const auto& resp) -> void {
```

Comment:
I think this is dangerous. This is an object that contains a shared_ptr to itself. This must ultimately end up with the function being destructed and destroying the share_ptr which would attempt to call the destructor again?